retention-period: Add archiving expired messages tool by retention pe…

kkanahin opened the pull request:
…riod

- Add models to store expired messages and expired user messages data after
  removing from original Message and UserMessage db tables.
- Add django migrations to create tables for archive models.
- Add tools for moving expired messages to archive tables.
- Add tools for deleting expired messages from Message and UserMessage tables.
- Add tests to check archiving and deliting tools.
- Add manage command to archive expired messages.

smarx commented:
*Automated message from Dropbox CLA bot*

@kkanahin, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on zerver/tests/test_retention.py:
```diff
@@ -26,89 +31,290 @@ def setUp(self):
+                ArchiveUserMessage.objects
+                    .filter(user_profile__realm=self.mit_realm)
+                    .order_by('id')
+                    .values_list('id', flat=True)
```
These tests are kinda long; might be worth writing some more test helper functions to simplify repetitive stuff.  

I feel like these tests use the pattern `list(foo.order_by('id').values_list('id', flat=True))` where `foo` is a query that maybe we should make the API here be e.g. `sorted_ids(ArchiveUserMessage.objects                    .filter(user_profile__realm=self.mit_realm))`

kkanahin replied:
sure will split this tests

timabbott commented:
Thanks for working on this @kkanahin!  I did a read through this, and it's looking really good; I posted one comment above on the tests; will review more closely later this week.

To what extent do you have outstanding TODOs for this?  The main things I notice are:
* We still need to add the management command code for actually deleting old ArchivedFoo stuff, with support for a retention duration (would just make it a server setting in the zproject/settings.py defaults at 30 days or something; see docs/settings.md for details) of the ArchivedFoo tables.
* We need to add ArchivedAttachment support (using `Attachment.Messages` to track whether the last message touching an attachment has been archived).

It's fine to work on those as added commits on top, though.

kkanahin commented:
@timabbott Your TODO items are correct and will be done in next iteration.

rishig commented:
Should we use the Django ORM for the SQL queries here?

kkanahin commented:
Django orm doesn't support moving  from one table to other in one query. It recieves all  data then convert them to list (I mean bulk_create) and insert to the new table. In my  methods it is done in one query withouth python processing.

timabbott commented:
For this operation, which is probably going to be our single most expensive type of SQL operation that Zulip does, I think it totally makes sense to do it in SQL.

rishig commented:
Got it, thanks!

timabbott commented on zerver/tests/test_retention.py:
```diff
@@ -324,3 +328,37 @@ def test_archiving_attachments_with_removing(self):
+        self.assertEqual(
+            ArchiveAttachment.objects.count(),
+            0
+        )
```
I don't see where this verifies that the attachments are actually deleted?

kkanahin replied:
It is checked do not deleting archive attachments if exists some  messages(not archived) which is  linked with this archive attachments. And deleting them if all related messages are archived and can be removed.

timabbott commented on zproject/settings.py:
```diff
@@ -1075,3 +1075,4 @@ def get_secret(key):
 PROFILE_ALL_REQUESTS = False
 
 CROSS_REALM_BOT_EMAILS = set(('feedback@zulip.com', 'notification-bot@zulip.com'))
+ARCHIVED_DATA_RETENTION_DAYS = 30
```
This needs to be user-configurable and thus should be added to DEFAULT_SETTINGS rather than set unconditionally here; see http://zulip.readthedocs.io/en/latest/settings.html for details on why.

timabbott commented on zerver/tests/test_retention.py:
```diff
@@ -251,3 +288,39 @@ def test_archive_message_tool(self):
         self._check_archive_messages_ids_by_realm(
             exp_msgs_ids_dict['mit_msgs_ids'], self.mit_realm)
 
+    def test_moving_attachments_to_archive_withouth_removing(self):
```
this test name is mispelled

kkanahin replied:
fixed

timabbott commented on zerver/lib/retention.py:
```diff
@@ -71,9 +113,20 @@ def delete_expired_user_messages():
+    # type: () -> None
+    removing_attachments = Attachment.objects.filter(
+        messages__isnull=True, id__in=ArchiveAttachment.objects.all())
+    removing_attachments.delete()
```
Call this `attachments_to_remove`.

kkanahin replied:
Fixed

kkanahin commented on puppet/zulip/files/cron.d/remove_old_archive_data:
```diff
@@ -0,0 +1,4 @@
+MAILTO=root
+
+# Remove old archive data . Time is in UTC.
+0 2 * * *   zulip cd /home/zulip/deployments/current && python manage.py remove_old_archive_data
```
Fixed

kkanahin commented on zerver/lib/retention.py:
```diff
@@ -130,3 +133,25 @@ def archive_messages():
+    arc_attachments.delete()
+
+
+def delete_archived_data():
```
Fixed

kkanahin commented:
@timabbott please review this pull request. I've added documentation page for retention feature and need your opinion about it. And could you review my answer for attachment test case.
Also I've added front-end part for retention period in admin settings.

timabbott commented:
The document could use a pass for English grammar, but otherwise looks good; I can get another contributor to tweak that.

I'll review the new code in detail in the next few days, but the one feature that I think this is still missing is a plan / tools for how we would unarchive some messages in the event that we discovered there had been a bug in this logic; it seems likely we'll eventually want to do that, and I'm thinking it might be wise to write the  code+tests for it now rather than when we have incident.

kkanahin commented:
I've fixed English grammar. Regarding restoring tool I will develop it.

kkanahin commented:
PR was updated. I've added restoring tools, console command and documentation part.

timabbott commented:
Can you rebase this onto master?

timabbott commented:
@kkanahin can you rebase this on top of master?

timabbott commented on docs/retention.md:
```diff
@@ -0,0 +1,114 @@
+# Retention policy
```
This should be linked to from docs/index.rst.

kkanahin replied:
@timabbott, Done

kkanahin commented:
sure @timabbott I'm merging

kkanahin commented:
@timabbott, merged.

kkanahin commented:
@timabbott PR has been updated and ready for review.

brainwane commented on puppet/zulip/files/cron.d/archive-messages:
```diff
@@ -0,0 +1,10 @@
+MAILTO=root
+# Removing old archived data should follow after archiving messages,
+# cause it checks related archived messages and original attachments rows before deleting .
```
Better phrasing, if I understand correctly:

~~~
# The retention policy data deletion process should remove old archived
# data *after* archiving messages, to check related archived messages
# and the original attachments' rows before deleting.
~~~

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+        return msgs
 
-    def _make_mit_messages(self, message_quantity, pub_date):
+    def _make_mit_msgs(self, msg_qauntity, pub_date):
```
This variable name should be `msg_quantity` wherever it's used.

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
-    def test_expired_messages_result_type(self):
+    def _send_cross_realm_message(self):
+        # type: () -> int
+        # Send message from bot to users from different realm.
```
I think this is singular, based on the function name, so the comment should be:

`# Send message from bot to user in a different realm.`

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+        archived_user_messages = ArchiveUserMessage.objects.all()
+        self.assertEqual(ArchiveMessage.objects.count(), 5)
+        self.assertEqual(archived_user_messages.count(), 10)
+        # Compare expected messages ids with archived messages in mit realm
```
I think this should be:

`# Compare expected messages' IDs with IDs of archived messages in MIT realm`

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+        self.assertEqual(archived_user_messages.count(), 10)
+        # Compare expected messages ids with archived messages in mit realm
+        self._check_archive_data_by_realm(exp_mit_msgs, self.mit_realm)
+        # Check no archive messages for zulip realm.
```
Does this mean checking messages for which the policy is not to archive, or checking that there exist zero messages to archive, or checking that zero messages have been archived? Please elaborate. :)

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+
+    def test_cross_realm_messages_archiving_one_realm_expired(self):
+        # type: () -> None
+        # Check archiving messages which is sent to different realms
```
This is singular, so, "Check archiving a message which is sent".

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+    def test_archive_message_tool(self):
+        # type: () -> None
+        # Check archiving tool.
+        exp_msgs_ids_dict = self._make_expired_messages()  # List of expired messages ids.
```
`List of expired messages' IDs.`

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+        exp_msgs_ids_dict = self._make_expired_messages()  # List of expired messages ids.
+        sended_cross_realm_message_id = self._send_cross_realm_message()
+        exp_msgs_ids_dict['mit_msgs_ids'].append(sended_cross_realm_message_id)
+        # Add cross realm message id.
```
`Add cross-realm message ID.`

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+
+    def test_cross_realm_messages_archiving_two_realm_expired(self):
+        # type: () -> None
+        # Check archiving cross realm message wich is expired on both realms.
```
I think this should be "cross-realm" with a hyphen.

brainwane commented on zerver/tests/test_retention.py:
```diff
@@ -21,95 +37,452 @@ class TestRetentionLib(ZulipTestCase):
+            timezone.now() - timedelta(days=101)
         )
+        exp_msgs_ids = exp_msgs_ids_dict['mit_msgs_ids'] + exp_msgs_ids_dict['zulip_msgs_ids']
+        # Get expired user messages by message ids
```
`Get expired user messages by message IDs.`

brainwane commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -89,6 +89,17 @@
+      <label for="realm_message_retention_days"
+        id="id_realm_message_retention_days_label"
+        title="{{t 'If empty, messages are not moved to archive.' }}">
+        {{t 'Retention period for messages in days (empty for disable)' }}
```
I'd like this to explicitly say something about data deletion, and be clearer about what "moved to archive" and "to disable" mean here, and linking to the help document (`retention`) would be good. I can take a crack at the phrasing after I edit `retention.md`.

brainwane commented:
@kkanahin In commit a7bdc6ad439406f01c5c9d435180af129af9ac2f please correct the spelling of "deleting" in your commit message.

Could you please add "Fixes: 106." to the end of one of your commit messages? (I might be wrong and this might only be a step towards fixing #106 -- let me know if I am.)

brainwane commented on zerver/management/commands/restore_messages_by_realm.py:
```diff
@@ -0,0 +1,21 @@
+    def add_arguments(self, parser):
+        # type: (ArgumentParser) -> None
+        parser.add_argument('domain', metavar='<domain>', type=str,
+                            help='domain of realm to restore messages data')
```
Suggested improvement: "The domain of the realm to restore messages data to."

brainwane commented:
@kkanahin I was looking through this in order to improve the documentation, and I'm a little confused and hope you can help me understand something.

[In the discussion in #106](https://github.com/zulip/zulip/issues/106#issuecomment-254110742) @timabbott mentioned that

> I'd like to use a model where after the N days the messages are deleted from Message, but added to a new table e.g. ArchivedMessage, and kept there for a week before being deleted finally, so that we have time to catch bugs for data is lost permanently.

Does your implementation do this? I looked at `retention.py` and I see where the archiving steps use the time delta between the current date and the publication date of the message and compare that against the number of days specified in the retention policy, but I don't think I see where the deletion steps wait the additional week after that. If I missed it, you should probably add some code comments or rename some variables to make them pop out better. Thanks!

brainwane commented:
@kkanahin I started on an improvement of `retention.md` in https://github.com/brainwane/zulip/commit/d70f8630769c0588f2ed2a0703110370265a4dc8 but I'm confused about the difference between 
`delete_expired_messages` and `delete_expired_user_messages`,  and the difference between those kinds of methods and the `delete_expired_archived_attachments`/`delete_expired_archived_data` set. Your documentation says "Records are removed from the actual tables when they don't have related objects", but should we be deleting anything from the main tables before moving it to the archived messages/attachments/data tables?

Thanks!

kkanahin commented:
thanks @brainwane for review. I will fix all bugs soon.

kkanahin commented:
> @kkanahin I was looking through this in order to improve the documentation, and I'm a little confused and hope you can help me understand something.
> 
> In the discussion in #106 @timabbott mentioned that
> 
> I'd like to use a model where after the N days the messages are deleted from Message, but added to a new table e.g. ArchivedMessage, and kept there for a week before being deleted finally, so that we have time to catch bugs for data is lost permanently.
> Does your implementation do this? I looked at retention.py and I see where the archiving steps use the time delta between the current date and the publication date of the message and compare that against the number of days specified in the retention policy, but I don't think I see where the deletion steps wait the additional week after that. If I missed it, you should probably add some code comments or rename some variables to make them pop out better. Thanks!

@brainwane I've added field "archived_date" to the ArchiveMessage, ArchiveUserMessage and ArchiveAttachment models as well as 'ARCHIVED_DATA_RETENTION_DAYS' configuration parameter to monitor archived messages life cycle.  You can find logic for removing archived messages in 'zerver.lib.retention.delete_expired_archived_data' function . For removing archived attachments I've created function 'zerver.lib.retention.delete_expired_archived_attachments'. I will add explanation comments to them.

kkanahin commented:
> @kkanahin I started on an improvement of retention.md in brainwane@d70f863 but I'm confused about the difference between
> delete_expired_messages and delete_expired_user_messages, and the difference between those kinds of methods and the delete_expired_archived_attachments/delete_expired_archived_data set. Your documentation says "Records are removed from the actual tables when they don't have related objects", but should we be deleting anything from the main tables before moving it to the archived messages/attachments/data tables?
> 
> Thanks!



> difference between
> delete_expired_messages and delete_expired_user_message
'delete_expired_user_message' - function for removing 'user_messages' (one message can belong to several users and this table is used for saving data about this)
'delete_expired_messages' - function for removing 'messages' from Message table after retention period. These messages can't be removed if one of a related user_messages is actual.



> and the difference between those kinds of methods and the delete_expired_archived_attachments/delete_expired_archived_data set

I've moved deleting expired archived attachments to separate function cause it requires more than one action(i.e removing files from storage) and it should be more comfortable to support this part.    

> Your documentation says "Records are removed from the actual tables when they don't have related objects", but should we be deleting anything from the main tables before moving it to the archived messages/attachments/data tables?

We shouldn't cause it's not safety. we must have a copy of data in archive tables before deleting from main tables. This logic allows avoiding accidental deletion data from main tables and data loss.

brainwane commented:
@showell or @timabbott I'm planning on doing an edit of `retention.md` to help improve this PR but I want to get a double-check from you that the architecture of the PR is what we want it to be, so I don't waste time documenting the wrong behavior. Thanks.

timabbott commented on docs/retention.md:
```diff
@@ -0,0 +1,140 @@
+Zulip contains console commands `management/commands/archive_messages.py`
+to move expired messages to archive tables and
+`management/commands/remove_old_archived_data.py` to remove archived
+messages and archived attachments permanently.
```
probably should mention these are expected to run in a cron job.

timabbott commented:
@brainwane yep, the architecture is correct and unlikely to change.

timabbott commented:
@kkanahin can you give this one a rebase?

timabbott commented:
A few more comments:
* The migrations need renumbering
* It'd be great if the management commands had a `--dry-run` option that would just report how many things would be archived/deleted, if that's easy to do without duplicating a lot of code.  I think one will definitely want those options to feel good about deploying this.

kkanahin commented:
Sure @timabbott. Sorry I've noticed your comments just today.

timabbott commented:
@lonerz @synicalsyntax this loop seems like a bug :)

lonerz commented:
@timabbot hmmm, I thought I took care of this bug already. Will look into it.

synicalsyntax commented:
Since @kkanahin mentioned the issue each time for his 8 commits, zulipbot posted 8 comments. @lonerz Are you working on this now or would you like me to open an issue for this?

lonerz commented:
@synicalsyntax you can open an issue and take care of it.

brainwane commented on docs/retention.md:
```diff
@@ -0,0 +1,140 @@
+# Retention policy
+
+
+## Overview
```
Most of the improvement suggestions I made in https://github.com/brainwane/zulip/commit/d70f8630769c0588f2ed2a0703110370265a4dc8 are still applicable to improve the clarity of this document - how about you integrate those (as long as I haven't introduced any inaccuracies) and then ping me so I can take a fresh look and help revise? Thanks!

timabbott commented:
@kkanahin this is failing tests; it looks like because you didn't add tests to `test_realm.py` for the new setting; can you fix that?  Also seems to have some mypy errors; didn't look at those though.

kkanahin commented:
@timabbott Yes will do it

timabbott commented:
@kkanahin I did some performance testing on the queries we're using, and I think unfortunately they're just way too slow to be able to use this at scale in production.

Some things I've confirmed (focusing on the `messages` query for now):
* If we can avoid joining against UserMessage, the query runs fairly fast (e.g. 50ms for a single medium-size realm with 1000 messages/day for a single historical day).
* If we do need to join against UserMessage, the query can take a really long time (e.g. > 2 minutes ever after limiting to just 1 day and just 1 medium-size realm), basically because it requires a sequential scan in the query plan (which is likely to be hopeless at scale).

So here's a proposed new architecture that would solve this problem:
(1) We do a loop over realms to shrink the commit sizes to be something manageable.  Later, if we offer per-stream limits, we can probably make this looping over `recipient_ids` with custom retention policies (and we'll need to do this anyway to handle the variability in retention policy).
(2) Within that loop, we first query all the messages in the expiry time window that were sent by a user in the realm (this will miss cross-realm messages, so we'll need to handle separately, but is at least fast!).  We can either return the "chopping block" messages to Python for further processing (since there shouldn't be that many), or proceed to create ArchivedMessage objects for them.
(3) Then, we created ArchivedUserMessage for any UserMessage entries in the realm that correspond to the new ArchivedMessage rows.  This should be fast since we'll have a fixed list of message IDs that shouldn't be too giant to translate here.
(4) Then, we move any MessageAttachment objects to the archive that go with a now-archived message.  
(5) Then, we check if any of the Attachment objects connected to any MessageAttachments that were moved via step (4) now have 0 messages; if so, those Attachment objects get moved to ArchivedAttachment.

This way, we only need to do the really expensive query against Message at the very beginning, and the rest we're just interacting with the subset that we're actually moving.

This misses some cross-realm messages, obviously.  We can handle cross-realm messages to streams pretty easily (just query all messages sent by the cross-realm bots with a recipient that's a stream in the realm); later, we can just create duplicate copies of cross-realm bots in each realm to effectively eliminate this whole issue of cross-realm stream messages.

The same works for single-recipient cross-realm PMs by cross-realm bots.

What remains is cross-realm group PMs, which are rare; I'm OK with the initial implementation ignoring them, and we can work on a specialized solution for that problem once it's resolved.  

Does that sound doable @kkanahin?  I realize this involves rewriting a bunch of code, but I think the result will be much more stable in the long term.

If I'm not mistaken, almost all of your tests just call `archive_messages` and check the state, so hopefully this just requires replacing the queries and not rewriting everything.

showell commented:
If we can have a set of small queries that identifies the message ids that need to be archived, then I would echo the idea of some kind of offline process that just handles them one by one.  Since realms are generally adding messages more quickly than removing messages (due to growth over time), it seems to be the offline process will always be able to keep up.  A refinement would be to run the archiver in "slow hours," although determining what those are could be difficult.

kkanahin commented:
@timabbott so, As I understood, Generally, we are going to split archiving by realm, sure it will increase speed. Maybe my way was wrong and didn't think about speed such as we are going to do it one time per day. Yes @timabbott it's quite doable, we just need to split queries by realm and rewrite some moments. So we can split data by realm on both sides(archive and actual tables).
Thank you, Tim for so detailed review. I got your idea.

timabbott commented:
Also @kkanahin I cleaned up and merged the setting piece (with the template section inside an "if false" so its inactive) just to avoid merge conflicts with other settings folks are working on.  That should save some potential future rebase pain :).

timabbott commented:
Also in the interest of minimizing carrying cost for this branch, I went ahead and merged the models and migrations changes, since I think we're pretty happy with those.

I made one small change while merging, though, which is renaming the `archived_date` fields to `archive_timestamp`, basically because we're planning to rename `pub_date` to just `timestamp`, and I think this will be more symmetric.


So when you next rebase:
* You can drop your frontend changes commit entirely
* You'll need to deal with the merge conflicts from models.py changes.

Thanks @kkanahin!  I hope this will help make life easier as we continue to get this working just how we want it to :).

kkanahin commented:
@timabbott I've updated this PR. Could you please review it. I went through all commits and split archiving by realm. Had a lot of problems with applying changes for each commit and I hope it is what you need. Also, I've improved all queries.

kkanahin commented:
I've checked queries performance too on more than 100000 rows, and it was good enough about 2 seconds per query. I've used 'hdd' for database.

timabbott commented:
Cool.  @kkanahin can you do a rebase of this?  I think I'm ready to merge the first several commits once that's done.  I think the merge conflicts result from b2504084ab306c0443b782a0c4e273367917f0ae.

kkanahin commented:
@timabbott, I've rebased this PR.

timabbott commented:
Now that we've merged the message deletion commit, can you rebase this to integrate with those changes @kkanahin?  Thanks!

kkanahin commented:
@timabbott sure

kkanahin commented:
@timabbott rebased

timabbott commented:
This PR is mostly complete, but hasn't been integrated due to insufficient time for careful testing.  It wouldn't be hard to pick up.

timabbott commented:
Closing in favor of the current version https://github.com/zulip/zulip/pull/12356.
